msgpack/_unpacker.pyx

```diff
@@ -129,12 +131,27 @@ def unpackb(object packed, object object_hook=None, object list_hook=None,
     cdef Py_ssize_t off = 0
     cdef int ret
 
-    cdef char* buf
+    cdef Py_buffer view
+    cdef char* buf = NULL
     cdef Py_ssize_t buf_len
     cdef char* cenc = NULL
     cdef char* cerr = NULL
-
-    PyObject_AsReadBuffer(packed, <const void**>&buf, &buf_len)
+    cdef char buffer_supported = 0
+
+    if PyObject_CheckBuffer(packed):
+        buffer_supported = 1
+        if PyObject_GetBuffer(packed, &view, PyBUF_SIMPLE) == 0:
+            if view.itemsize != 1:
+                PyBuffer_Release(&view)
+                raise ValueError("cannot unpack from multi-byte object")
+            buf_len = view.len
+            buf = <char*> view.buf
+    else:
+        PyObject_AsReadBuffer(packed, <const void**>&buf, &buf_len)
+        PyErr_WarnEx(DeprecationWarning,
+                     "unpacking %s requires old buffer protocol,"
+                     "which will be removed in msgpack 1.0" % type(packed),
```

Comment:
I don't know if there is any convention, but if the message is a whole sentence it does feel right to add proper punctuation.